### PR TITLE
Rework node copy task

### PR DESCRIPTION
**Pull request by ray6080**
This PR reworks node copy task to be consistent with the logic of processor task. Also, unifies csv, parquet and npy to go through the same task system.

### CODE REVIEW

File: src/storage/copier/node_copy_executor.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    case common::CopyDescription::FileType::NPY: {
+        for (auto i = 0u; i < copyDescription.filePaths.size(); i++) {
+            std::vector<std::string> filePaths(1);
+            filePaths[0] = copyDescription.filePaths[i];
+            std::shared_ptr<NodeCopySharedState> sharedState =
+                std::make_shared<NodeCopySharedState>(filePaths, fileBlockInfos);
+            auto nodeCopier = std::make_unique<NPYNodeCopier<T>>(std::move(sharedState),
+                pkIndex.get(), copyDescription, columns, i,
+                i == propertyIDToColumnIDMap[primaryKey.propertyID] ? 0 : INVALID_COLUMN_ID);
+            tasks.push_back(
```
Comment (andyfengHKU) [inline src/storage/copier/node_copy_executor.cpp:78]: you can probably pull task.push_back out of the case switch

File: src/include/storage/copier/node_copier.h [hunk 2]
```diff
@@ -1,136 +1,235 @@
...
+        common::offset_t offset, HashIndexBuilder<T>* pkIndex) {
+        assert(false);
+    }
 
-    std::unordered_map<common::property_id_t, std::unique_ptr<InMemNodeColumn>> columns;
+protected:
+    std::shared_ptr<NodeCopySharedState> sharedState;
+    HashIndexBuilder<T>* pkIndex;
+    common::CopyDescription copyDesc;
+    std::vector<std::unique_ptr<InMemNodeColumn>>& columns;
```
Comment (andyfengHKU) [inline src/include/storage/copier/node_copier.h:131]: personally I usually use std::vector<InMemNodeColumn*> and not referencing the vector

File: src/include/storage/copier/node_copier.h [hunk 3]
```diff
@@ -1,136 +1,235 @@
...
+class CSVNodeCopier : public NodeCopier<T> {
+public:
+    CSVNodeCopier(std::shared_ptr<NodeCopySharedState> sharedState, HashIndexBuilder<T>* pkIndex,
+        const common::CopyDescription& copyDesc,
+        std::vector<std::unique_ptr<InMemNodeColumn>>& columns, common::column_id_t pkColumnID)
+        : NodeCopier<T>{std::move(sharedState), pkIndex, copyDesc, columns, pkColumnID} {}
+
+    inline std::unique_ptr<NodeCopier<T>> clone() const override {
+        return std::make_unique<CSVNodeCopier<T>>(
+            this->sharedState, this->pkIndex, this->copyDesc, this->columns, this->pkColumnID);
```
Comment (andyfengHKU) [inline src/include/storage/copier/node_copier.h:154]: remove `this->` and for other copiers.

File: src/include/storage/copier/node_copier.h [hunk 4]
```diff
@@ -1,136 +1,235 @@
...
+          csvReaderConfig{csvReaderConfig}, tableSchema{tableSchema} {};
+
     std::unique_ptr<NodeCopyMorsel> getMorsel() override;
 
 private:
-    std::shared_ptr<arrow::csv::StreamingReader> csvStreamingReader;
+    common::CSVReaderConfig* csvReaderConfig;
+    catalog::TableSchema* tableSchema;
+    std::shared_ptr<arrow::csv::StreamingReader> reader;
+    std::string currentFilePath;
```
Comment (andyfengHKU) [inline src/include/storage/copier/node_copier.h:88]: track fileIdx instead of path

**Review by andyfengHKU — APPROVED** [review]

File: src/include/storage/copier/node_copier.h (hunk 3 above)
Comment (ray6080) [reply src/include/storage/copier/node_copier.h:154]: `this` is to access base class fields here.

File: src/storage/copier/node_copy_executor.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        auto fName = StorageUtils::getNodePropertyColumnFName(
+            outputDirectory, tableSchema->tableID, property.propertyID, DBFileType::WAL_VERSION);
+        columns.push_back(
+            NodeInMemColumnFactory::getNodeInMemColumn(fName, property.dataType, numRows));
+        propertyIDToColumnIDMap[property.propertyID] = columns.size() - 1;
+    }
+    logger->info("Done initializing in memory columns.");
+}
+
+void NodeCopyExecutor::populateColumnsAndLists(processor::ExecutionContext* executionContext) {
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/storage/copier/node_copy_executor.cpp:25]: It is specific to NodeCopyExecutor. I think #971 is the issue.

File: src/storage/copier/node_copy_executor.cpp (hunk 1 above)
Comment (ray6080) [reply src/storage/copier/node_copy_executor.cpp:78]: Not really, npy is specially handled right now, as each file is assigned a task.

File: src/include/storage/copier/node_copier.h (hunk 4 above)
Comment (ray6080) [reply src/include/storage/copier/node_copier.h:88]: Good catch. Actually I don' think I need to track this. Removed now.

**Pull request merged**